x-pack/plugins/lens/public/indexpattern_datasource/operations/definitions/filters/filter_popover.tsx

```diff
@@ -0,0 +1,196 @@
+ */
+import './filter_popover.scss';
+
+import React, { MouseEventHandler, useState } from 'react';
+import { EuiPopover, EuiFieldText, EuiSpacer, keys } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { debounce } from 'lodash';
+import { FilterValue, defaultLabel, isQueryValid } from '.';
+import { IndexPattern } from '../../../types';
+import { QueryStringInput, Query } from '../../../../../../../../src/plugins/data/public';
+
+export const FilterPopover = ({
+  filter,
+  setFilter,
+  indexPattern,
+  Button,
+  isOpenByCreation,
+  setIsOpenByCreation,
+}: {
+  filter: FilterValue;
+  setFilter: Function;
+  indexPattern: IndexPattern;
+  Button: React.FunctionComponent<{ onClick: MouseEventHandler }>;
+  isOpenByCreation: boolean;
+  setIsOpenByCreation: Function;
+}) => {
+  const [isPopoverOpen, setIsPopoverOpen] = useState(false);
+  const inputRef = React.useRef<HTMLInputElement>();
+
+  const setPopoverOpen = (isOpen: boolean) => {
+    setIsPopoverOpen(isOpen);
+    setIsOpenByCreation(isOpen);
+  };
+
+  const setFilterLabel = (label: string) => setFilter({ ...filter, label });
+  const setFilterQuery = (input: Query) => setFilter({ ...filter, input });
+
+  const getPlaceholder = (query: Query['query']) => {
+    if (query === '') {
+      return defaultLabel;
+    }
+    if (query === 'object') return JSON.stringify(query);
+    else {
+      return String(query);
+    }
+  };
+
+  return (
+    <EuiPopover
+      anchorClassName="eui-fullWidth"
+      panelClassName="lnsIndexPatternDimensionEditor__filtersEditor"
+      isOpen={isOpenByCreation || isPopoverOpen}
+      ownFocus
+      closePopover={() => {
+        setPopoverOpen(false);
+      }}
+      button={
+        <Button
+          onClick={() => {
+            setIsPopoverOpen((open) => !open);
+            setIsOpenByCreation(false);
+          }}
+        />
+      }
+    >
+      <QueryInput
+        isInvalid={!isQueryValid(filter.input, indexPattern)}
+        value={filter.input}
+        indexPattern={indexPattern}
+        onChange={setFilterQuery}
+        onSubmit={() => {
+          if (inputRef.current) inputRef.current.focus();
+        }}
+      />
+      <EuiSpacer size="s" />
+      <LabelInput
+        value={filter.label || ''}
+        onChange={setFilterLabel}
+        placeholder={getPlaceholder(filter.input.query)}
+        inputRef={inputRef}
+        onSubmit={() => setPopoverOpen(false)}
+      />
+    </EuiPopover>
+  );
+};
+
+const QueryInput = ({
+  value,
+  onChange,
+  indexPattern,
+  isInvalid,
+  onSubmit,
+}: {
+  value: Query;
+  onChange: (input: Query) => void;
+  indexPattern: IndexPattern;
+  isInvalid: boolean;
+  onSubmit: () => void;
+}) => {
+  const [inputValue, setInputValue] = useState(value);
+
+  React.useEffect(() => {
+    setInputValue(value);
+  }, [value, setInputValue]);
+
+  const onChangeDebounced = React.useMemo(() => debounce(onChange, 256), [onChange]);
```

Comment:
These warnings happen in most of places when we modify the state so deeply. I think We need more top-down approach here instead of just trying to patch this single place.